src/mo_interpreter/interpret.ml

```diff
@@ -885,16 +891,22 @@ and match_shared_pat env shared_pat c =
 
 (* Objects *)
 
-and interpret_obj env obj_sort dec_fields (k : V.value V.cont) =
+and interpret_obj env obj_sort self_id dec_fields (k : V.value V.cont) =
   match obj_sort with
   | T.Actor ->
      let self = V.fresh_id() in
+     let self' = V.Blob self in
+     (* Define self_id eagerly *)
+     let env' = match self_id with
+     | Some id -> adjoin_vals env (declare_id id)
+     | None -> env in
+     Option.iter (fun id -> define_id env' id self') self_id;
      let ve_ex, ve_in = declare_dec_fields dec_fields V.Env.empty V.Env.empty in
-     let env' = adjoin_vals { env with self = self } ve_in in
-     interpret_dec_fields env' dec_fields ve_ex
+     let env'' = adjoin_vals { env' with self } ve_in in
+     interpret_dec_fields env'' dec_fields ve_ex
      (fun obj ->
         (env.actor_env := V.Env.add self obj !(env.actor_env);
```

Comment:
Crazy idea: At the moment, the actor environment maps actor ids to Objects containing a map from field names to values.
Suppose we changed that, so an actor environment just contains a map from field names to promises (i.e. `ve_ex`).
Then we can update the actor map before we interpret the body. As the promises in `ve_ex` get defined, so do the ones accessible from the self-reference, a bit like your previous 'increments' but automatic.
I think that might let us implement the equality on mixtures of  symbolic and function references (provided we also give the equality operator) access to the environment. 

Best attempted in a separate PR I think. Might be a dead-end with too many pervasive changes.

Reply:
On first look this might work, the callback could simply define the method as `interpret_obj` is iterating over the decs.